### PR TITLE
Merge all package tracking IA's into 1 (Work in progress #1138 )

### PR DESCRIPTION
**Pull request by dronov**
This is a start for fixing #1138 
I've made code structure. I added just Hongkong Post, Post Norden and Russian Post without any assets. I'll upload it later. 

It could be great If someone gives me any feedback looking at my code. It based on HKDK and one another module and I'm interested to make it clear and readable.

UPD: of course this is not a final code. I has to be reviewed.

### CODE REVIEW

File: share/goodie/package_tracking/responses.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+HongKong:
+  name: "Hongkong Post"
+  tracking_url: http://app3.hongkongpost.com/CGI/mt/genresult.jsp?tracknbr=
+  short_code: hk$
```
Comment (moollaza) [inline share/goodie/package_tracking/responses.yml]: nit: flip this and the previous line for consistency :)

**Comment by moollaza** [comment]
@dronov thanks a lot for tackling this! We really apprecaite it :smile: 

This is definitely a step in the right direction. I think you'll want to adjust the Perl so that you collect all the tracking companies that succesfully matched the given code and then if we had 2+ matches, you'll want to adjust the template you're going to use.

Also, this should now be updated to use the Goodie templates. We *just* launched some documention outlining how this works for Goodies: https://duck.co/duckduckhack/goodie_displaying

If we have 2+ matches you'll want to set the template group to `list` otherwise it'll need to be the icon template and the icons etc will have to be set accordingly.

**Comment by dronov** [comment]
@moollaza thank you for feedback! I'm glad that I had chosen the right way to tackle this :) 

This is an interesting task cause many postal services have a unique tracking codes. But currently I;m working on use cases with one service matches with user's query (e.g. I've added the Lasership tracking). When I'll meet use cases with 2+ matches tracking companies I'll change the main logic of code and make the correct layout.

**Comment by moollaza** [comment]
Sounds good! Let us know when you're ready for more review :+1:

**Comment by dronov** [comment]
@moollaza I didn't forget about my PR. I had not a lot of free time on last week but I'll back to PR tomorrow.

**Comment by dronov** [comment]
Hello again, @moollaza! 

I need an advice. As I looked at the code again and again I realized that I need a little bit complicated solution for detecting is that the users query a real postal service or not. Many of classes have an different algorythms to check matched user's query for correct tracking number (e.g. Fedex, DHL, UPS, USPS). I think that writing one big perl package named PackageTracking wouldn't be the good solution. Based on this fact I think that the implementing of Strategy Pattern would be better: we have some classes (packages) for every Postal Service with methods to decide is that the user's query a real tracking code or not. In base strategy we could handle the common words like 'tracking, track' and codes like 'XX12356789YY'. After that we could create a concrete Strategy of some Postal Service which checks correctness of the code.

What do you think about this? If my thoughts make a sense could I make some directory like lib/DDG/Goodie/PackageTracking to implement the Strategy Pattern and hold necessary packages there?

PS. I thought about the simpler solutions, but all of them had a lot of code with difficult readable structure.

**Comment by moollaza** [comment]
@dronov sorry for the delay! Let me think on this a bit and get back to you!

**Comment by dronov** [comment]
@moollaza ping :) 
I still would like to get an advice...

**Comment by moollaza** [comment]
@dronov very sorry for the delay!

I'm not familiar with the Strategy Pattern, but your approach sounds reasonable. I had hoped that the checks required to determine which postal service the input belongs to could be greatly simplified (ideally just a pattern match), so that we wouldn't need to have several modules.

Such is not that case ;) so we'll have to handle things a little differently. Your suggested approach means that the IA will hopefully be extensible, so new sub-modules can be added for new postal services and ideally the main controller won't have to change much, or at all.

I'm confident whatever you arrive at will be a sound solution, so please go ahead and get started. Feel free to ping me if you need more feedback!

Thanks again for contributing @dronov, and sorry about the delay!

**Comment by mintsoft** [comment]
@dronov Just checking in, are you still interested in finishing this off?

**Pull request closed without merging**